### PR TITLE
MutationObserver api: Implement takeRecords() and characterData mutations

### PR DESCRIPTION
**Pull request by fabricedesre**
I started by looking at adding support for `takeRecords()` to get fluent.js to work on Servo. To get the wpt tests to pass I ended up implementing character data mutations (most of the plumbing was already there) and this fixed a few more tests.

We are still missing support for `disconnect()`, which I keep for a followup.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach build-geckolib` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20689)

### CODE REVIEW

File: components/script/dom/mutationobserver.rs
Comment (emilio) [inline components/script/dom/mutationobserver.rs]: can you rename this to `interested_observers` while at it, maybe in a different commit?

File: components/script/dom/mutationobserver.rs
Comment (emilio) [inline components/script/dom/mutationobserver.rs]: nit: You want `find` instead of `position` + the index.

File: components/script/dom/mutationobserver.rs
Comment (emilio) [inline components/script/dom/mutationobserver.rs]: The code duplication here is somewhat unfortunate, can we move the common code somewhere, maybe a local closure?

That being said, it's not much code... Your call I guess.

File: components/script/dom/mutationrecord.rs
Comment (emilio) [inline components/script/dom/mutationrecord.rs]: nit: Please indent it with block indentation:

```rust
pub fn character_data_mutated(
    target: &Node,
    old_value: Option<DOMString>,
) -> DomRoot<MutationRecord> {
```

File: components/script/dom/characterdata.rs
Comment (emilio) [inline components/script/dom/characterdata.rs]: A bit unfortunate, this is probably a tad slower...

We had `content_changed`, which we can probably inline now in ReplaceData... But maybe it's worth to do both `content_will_change` and `content_changed`, and keep the fast path here, triggering the mutation in `content_will_change`.

**Review by emilio — COMMENTED** [review]
Looks fine to me, mod nits and the ReplaceData bit which may be a bit concerning...

File: components/script/dom/mutationobserver.rs
Comment (fabricedesre) [reply components/script/dom/mutationobserver.rs]: done.

File: components/script/dom/mutationobserver.rs
Comment (fabricedesre) [reply components/script/dom/mutationobserver.rs]: Using find() leads to a bunch of problems with needing to borrow mutably interested_observer twice, and working around that is actually more code than currently and hurts readability quite a lot.
So unless I'm missing a trick, I'd rather keep it like that.

File: components/script/dom/mutationrecord.rs
Comment (fabricedesre) [reply components/script/dom/mutationrecord.rs]: I followed the local style in this file.

File: components/script/dom/characterdata.rs
Comment (fabricedesre) [reply components/script/dom/characterdata.rs]: I added a `queue_mutation_record` method that is called from both `SetData` and `ReplaceData`.

File: components/script/dom/mutationrecord.rs
Comment (KiChjang) [reply components/script/dom/mutationrecord.rs]: Some of our files are still using non-standard Rust styling, since they were merged before we had proper styling through rustfmt, and we're refactoring them as we spot them.

Comment (fabricedesre) [reply components/script/dom/mutationrecord.rs]: And in these cases you refactor the whole file I guess?

**Comment by fabricedesre** [comment]
friendly ping @emilio

File: components/script/dom/mutationobserver.rs
Comment (emilio) [inline components/script/dom/mutationobserver.rs]: nit: indentation.

File: components/script/dom/mutationobserver.rs
Comment (emilio) [inline components/script/dom/mutationobserver.rs]: nit: indentation

File: components/script/dom/mutationrecord.rs
Comment (emilio) [reply components/script/dom/mutationrecord.rs]: not necessarily... Feel free to fix it or not though.

**Review by emilio — APPROVED** [review]

File: tests/wpt/metadata/dom/nodes/MutationObserver-takeRecords.html.ini
Comment (emilio) [inline tests/wpt/metadata/dom/nodes/MutationObserver-takeRecords.html.ini]: Oh, this file can completely be removed now, btw.

**Comment by emilio** [comment]
@bors-servo r-

Let's make sure this lands with the correct reviewer name. If you get to remove the file before it merges, please use `@bors-servo r=emilio` instead of just `r+`, since otherwise the commit will land self-reviewed.

**Pull request merged**
